pkg/apiserver/certificate/certificate.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certificate
+
+import (
+	"fmt"
+	"net"
+
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	"k8s.io/apiserver/pkg/server/options"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+)
+
+var (
+	// certDir is the directory that the TLS Secret should be mounted to. Declaring it as variable for testing.
+	certDir = "/var/run/antrea/antrea-controller-tls"
+	// The DNS names that the TLS certificate will be signed with.
+	// Note that antrea-agent and kube-aggregator only verify the short server name.
+	// The second server name may not work if user specify a different cluster domain.
+	AntreaServerNames = []string{
+		"antrea.kube-system.svc",
+		"antrea.kube-system.svc.cluster.local",
+	}
+)
+
+func ApplyServerCert(selfSignedCert bool, client kubernetes.Interface, aggregatorClient clientset.Interface, secureServing *options.SecureServingOptionsWithLoopback) (*CACertController, error) {
+	var err error
+	var caContentProvider dynamiccertificates.CAContentProvider
+	if selfSignedCert {
+		// Set the PairName but leave certificate directory blank to generate in-memory by default.
+		secureServing.ServerCert.CertDirectory = ""
+		secureServing.ServerCert.PairName = "antrea-controller"
+
+		if err := secureServing.MaybeDefaultWithSelfSignedCerts("antrea-controller", AntreaServerNames, []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
+			return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
+		}
+
+		certPEMBlock, _ := secureServing.ServerCert.GeneratedCert.CurrentCertKeyContent()
+		caContentProvider, err = dynamiccertificates.NewStaticCAContent("self-signed cert", certPEMBlock)
+		if err != nil {
+			return nil, fmt.Errorf("error reading self-signed CA certificate: %v", err)
+		}
+	} else {
+		// Since 1.17.0 (https://github.com/kubernetes/kubernetes/commit/3f5fbfbfac281f40c11de2f57d58cc332affc37b),
+		// apiserver reloads certificate cert and key file from disk every minute, allowing serving tls config to be updated.
+		certFile := certDir + "/tls.crt"
+		keyFile := certDir + "/tls.key"
```

Comment:
maybe using `path.Join` (https://golang.org/pkg/path/) is a cleaner way of doing this?

Reply:
Done